target/ppc/int_helper.c

```diff
@@ -910,6 +910,79 @@ VCT(uxs, cvtsduw, u32)
 VCT(sxs, cvtsdsw, s32)
 #undef VCT
 
+#define XVIGER_NAME(bits, rank)                                         \
+                    glue(glue(helper_xvi, bits), glue(ger, rank))
+
+#define VsrN(i) VsrB((i >> 1)) >> ((i+1) % 2)*4 & 0xF
+#define VsrSN(i) VsrB((i >> 1)) >> ((i+1) % 2)*4 & 0xF | (0xF ^ (-1))
+
+#define EXTS(num, nb)                                                   \
+             (((((num) >> (nb - 1)) & 1) == 0) ?                        \
+             num :                                                      \
+             ((((1 << nb) - 1) ^ (-1)) | (num)))
+
+#define XVIGER(bits, rank, el)                                          \
+    void XVIGER_NAME(bits, rank) (CPUPPCState *env,                     \
+                                uint32_t a_r, uint32_t b_r,             \
+                                uint32_t  at_r, uint64_t mask)          \
+    {                                                                   \
+        uint8_t pmsk, xmsk, ymsk, opmsk;                                \
+        ppc_vsr_t *a, *b, *at;                                          \
+        int64_t psum;                                                   \
+        int32_t va, vb;                                                 \
+        int i, j, k;                                                    \
+        bool sat;                                                       \
+        xmsk = (uint8_t)mask & 0x0F;                                    \
+        ymsk = (uint8_t)(mask >> 4) & 0x0F;                             \
+        pmsk = (uint8_t)(mask >> 8) & 0xFF;                             \
+        opmsk = (mask >> 32) & 0xFF;                                    \
+        a = cpu_vsr_ptr(env, a_r);                                      \
+        b = cpu_vsr_ptr(env, b_r);                                      \
+        at = cpu_vsr_ptr(env, at_r);                                    \
+        sat = (bool)(opmsk & SAT_MASK);                                 \
+        for (i = 0; i < 4; i++) {                                       \
+            for (j = 0; j < 4; j++) {                                   \
+                if ((1 << (3-i) & xmsk) && (1 << (3-j) & ymsk)) {       \
```

Comment:
Changed it, but @mferst suggested maybe moving the xmsk "if" to outside the second "for", because if that bit is 0 then the entire line should be set to 0

Reply:
Good point!

Reply:
When you skip the whole line, don't forget to set all of the elements in that line to zero